plugins/hook/hook.7
.TH YED-HOOK 7 "YED Plugin Manuals" "" "YED Plugin Manuals"
.SH NAME
hook \- Run commands on editor events.
.SH CONFIGURATION
None
.SH COMMANDS
.SS hook [options] <event> <command>
Run <command> when <event> fires.
.SS hook-list-options
Print the available options to *log.
.SS hook-list-events
Print the available events to the command line.
.SH BUFFERS
None
.SH NOTES
This plugin is just a proof of concept at this point, but should be developed out more eventually.
.SH VERSION
0.0.1
.SH KEYWORDS
when, event, hook, pre, post, command, callback, handler


CHANGELOG.md
# Changelog

## 1306 - 2021-12-13
### Fixed
    - fixed a bug that would cause command line contents to overflow into a new terminal line.

## 1305 - 2021-12-13
### Fixed
    - fixed a bug that would cause some cursor movements to crash on binary buffer contents.

## 1304 - 2021-12-10
### Added
    - new command `frame` activates frame based on their number.

## 1303 - 2021-12-8
### Changed
    - the `plugins` directory is now a `subtree`. This makes installing simpler and allows you to just grab a `tar` of a release without needing `git`.

## 1302 - 2021-12-7
### Changed
    - removed use of `codesign` in the install script for Darwin targets. This isn't needed any more and will help us package `yed`.

## 1301 - 2021-12-7
### Fixed
    - Fixed logic in yed_cell_is_in_frame()

## 1300 - 2021-12-6
### Added
- Started keeping a changelog.
- Started tagging versions in git.
- A new syntax highlighting engine in `syntax.h`.
- Status line additions:
    - `%{...}` syntax for expanding a variable that contains an attribute string
    - Padding and justification on `%` expansions: `%` followed by an integer, `x`, will be padded to `x` characters. If `x` is preceded by a `-`, the text will be left justified. A `=` indicates center justification. The default is right justification. This has no effect on expansions that result in attributes.
    - `%T`: 24-hour clock
- Events:
    - `EVENT_STATUS_LINE_PRE_UPDATE` will trigger before the status line is about to be updated.
    - `EVENT_PLUGIN_MESSAGE` is a new general purpose event that plugins can send and other plugins can handle.
- Mouse support added to the core for left, center, and right buttons as well as up and down scroll wheel. Distinguishes between press, release, and drag.
- Useful functions for getting buffer contents as byte strings:
    - `char *yed_get_selection_text(yed_buffer *buffer);`
    - `char *yed_get_line_text(yed_buffer *buffer, int row);`
    - `char *yed_get_buffer_text(yed_buffer *buffer);`
- Style components:
    - `good`: general indication in the positive (e.g. a notification that a build succeeded)
    - `bad`: general indication in the negative (e.g. a notification that a build failed)
    - `popup`: the foreground and background to be used for popup menus
    - `popup-alt`: the foreground and background to be used for other popup menu elements (e.g. the selected item)
    - `code-variable`: for variables in code
    - `code-field`: for fields of a structure or class
    - `code-escape`: for escape sequences in string literals
### Changed
- `highlight.h` is now deprecated and will be removed in version 1400.
- Status line: `%T`, which used to expand to the filetype, now expands to a 24-hour clock. `%F` now expands to the filetype.
### Removed
- Removed the default command `if`.
### Fixed
- A buffer's undo history is now reset when `buffer-reload` gets run (or whenever the contents of the buffer are loaded from a file).
- Plugins are now install-agnostic -- they will use whatever `libyed.so` is being used by the binary. This fixes a crash that would occur when installations would mix on the same system.


hook/hook.c
#include <yed/plugin.h>

static yed_plugin *Self;

#include <yed/tree.h>
use_tree(int, array_t);

static tree(int, array_t) hooks_by_event;

typedef struct {
    char    *cmd;
    array_t  cmd_args;
} hook_t;

void unload(yed_plugin *self);
void hook_list_options(int n_args, char **args);
void hook_list_events(int n_args, char **args);
void hook(int n_args, char **args);

int yed_plugin_boot(yed_plugin *self) {
    YED_PLUG_VERSION_CHECK();

    Self = self;

    hooks_by_event = tree_make(int, array_t);

    yed_plugin_set_command(self, "hook",              hook);
    yed_plugin_set_command(self, "hook-list-options", hook_list_options);
    yed_plugin_set_command(self, "hook-list-events",  hook_list_events);

    yed_plugin_set_unload_fn(self, unload);

    return 0;
}

void free_hook(hook_t *h) {
    char **it;

    free(h->cmd);
    array_traverse(h->cmd_args, it) {
        free(*it);
    }
    array_free(h->cmd_args);
}

void unload(yed_plugin *self) {
    tree_it(int, array_t)  event_it;
    hook_t                *hook_it;

    tree_traverse(hooks_by_event, event_it) {
        array_traverse(tree_it_val(event_it), hook_it) {
            free_hook(hook_it);
        }

        array_free(tree_it_val(event_it));
    }

    tree_free(hooks_by_event);
}

void execute_hook(hook_t *h) {
    yed_execute_command(h->cmd,
                        array_len(h->cmd_args),
                        array_data(h->cmd_args));
}

void hook_handler(yed_event *event) {
    tree_it(int, array_t)  lookup;
    hook_t                *hook_it;


    if (!tree_it_good(lookup
        = tree_lookup(hooks_by_event, event->kind))) { return; }

    array_traverse(tree_it_val(lookup), hook_it) {
        execute_hook(hook_it);
    }
}

void add_hook(hook_t *h, int event) {
    tree_it(int, array_t) it;
    yed_event_handler     handler;

    it = tree_lookup(hooks_by_event, event);
    if (!tree_it_good(it)) {
        it = tree_insert(hooks_by_event, event, array_make(hook_t));
        handler.kind = event;
        handler.fn   = hook_handler;
        yed_plugin_add_event_handler(Self, handler);
    }

    array_push(tree_it_val(it), *h);
}

void hook_list_options(int n_args, char **args) {
}
int hook_arg(hook_t *h, char *arg) {
    return 1;
}


void hook_list_events(int n_args, char **args) {
    yed_cprint("\ncursor-moved");
    yed_cprint("\npost-write");
}
int hook_event(char *arg) {
#define CHECK_EV(str, enu) \
    if (strcmp(arg, (str)) == 0) { return (enu); }

    CHECK_EV("cursor-moved", EVENT_CURSOR_POST_MOVE);
    CHECK_EV("post-write",   EVENT_BUFFER_POST_WRITE);

    return -1;
#undef CHECK_EV
}

void hook(int n_args, char **args) {
    hook_t h;
    int    event;
    int    i;
    char * arg_dup;

    memset(&h, 0, sizeof(h));

    for (i = 0; i < n_args && strncmp(args[i], "--", 2) == 0; i += 1) {
        if (!hook_arg(&h, args[i])) { goto arg_error; }
    }

    if (n_args - i < 2) { goto arg_error; }

    if ((event = hook_event(args[i])) == -1) { goto arg_error; }

    i     += 1;
    h.cmd  = strdup(args[i]);

    if (i < n_args - 1) {
        h.cmd_args = array_make(char*);
        for (i = i + 1; i < n_args; i += 1) {
            arg_dup = strdup(args[i]);
            array_push(h.cmd_args, arg_dup);
        }
    }

    add_hook(&h, event);

    goto out;

arg_error:;
    yed_cerr("usage: hook [OPTIONS] EVENT CMD [CMD_ARGS...]\n"
             "    (run 'hook-list-options' or 'hook-list-events' for more info)");
out:;
    return;
}
